### PR TITLE
HistogramData rollout: QwtWorkspaceBinData.cpp to DPDFInputDataControl.cpp

### LINKED ISSUE #18793: HistogramData rollout: QwtWorkspaceBinData.cpp to DPDFInputDataControl.cpp
This issue is part of the HistogramData rollout, see #17641.
Algorithms covered in this issue are:

- MantidQt/API/src/QwtWorkspaceBinData.cpp
- MantidQt/API/src/QwtWorkspaceSpectrumData.cpp
- MantidQt/CustomDialogs/src/FitDialog.cpp
- MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/MultiDatasetFit/MDFDatasetPlotData.h
- MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Muon/ALCBaselineModellingView.h
- MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Muon/ALCDataLoadingView.h
- MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Muon/ALCPeakFittingView.h
- MantidQt/CustomInterfaces/src/DynamicPDF/DPDFDisplayControl.cpp
- MantidQt/CustomInterfaces/src/DynamicPDF/DPDFFourierTransform.cpp
- MantidQt/CustomInterfaces/src/DynamicPDF/DPDFInputDataControl.cpp

### PR DESCRIPTION
**Pull request by PranavBahuguna**
This PR is part of refactoring effort to use the new interface of the HistogramData module.
See #17641 for details.

The following algorithms were converted to use HistogramData:

- MantidQt/API/src/QwtWorkspaceBinData.cpp
- MantidQt/API/src/QwtWorkspaceSpectrumData.cpp
- MantidQt/CustomDialogs/src/FitDialog.cpp
- MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/MultiDatasetFit/MDFDatasetPlotData.cpp
- MantidQt/CustomInterfaces/src/DynamicPDF/DPDFFourierTransform.cpp
- MantidQt/CustomInterfaces/src/DynamicPDF/DPDFInputDataControl.cpp

No performance have been added.
No algorithms show significantly improved performance.

**To test:**

Mostly code review. Only `QwtWorkspaceBinData.cpp` and `QwtWorkspaceSpectrumData.cpp` have associated unit tests.

Fixes #18793.

### CODE REVIEW

File: MantidQt/API/test/QwtWorkspaceBinDataTest.h [hunk 1]
```diff
@@ -19,10 +19,10 @@ class QwtWorkspaceBinDataTest : public CxxTest::TestSuite {
     for (size_t i = 0; i < 3; i++) {
       ax1->setValue(i, 10.0 + static_cast<double>(i));
       for (size_t j = 0; j < 5; j++)
-        ws->dataX(i)[j] = double(i) + double(j);
+        ws->mutableX(i)[j] = double(i) + double(j);
```
Comment (LamarMoore) [inline MantidQt/API/test/QwtWorkspaceBinDataTest.h]: `HistogramData::LinearGenerator`? Shouldn't use non-const access in a tight loop (I know `i` only goes to 3 but it is better not to practice this things)

File: MantidQt/API/test/QwtWorkspaceBinDataTest.h [hunk 2]
```diff
@@ -19,10 +19,10 @@ class QwtWorkspaceBinDataTest : public CxxTest::TestSuite {
...
       for (size_t j = 0; j < 4; j++) {
-        ws->dataY(i)[j] = double(i) + double(j) * 2;
-        ws->dataE(i)[j] = double(i) + double(j) * 3;
+        ws->mutableY(i)[j] = double(i) + double(j) * 2;
```
Comment (LamarMoore) [inline MantidQt/API/test/QwtWorkspaceBinDataTest.h]: Same as above. `ws->setHistogram(i, BinEdges(size, LinearGenerator(i...`. Maybe?

File: MantidQt/API/test/QwtWorkspaceSpectrumDataTest.h [hunk 3]
```diff
@@ -16,10 +16,10 @@ class QwtWorkspaceSpectrumDataTest : public CxxTest::TestSuite {
     ws = WorkspaceCreationHelper::create2DWorkspace(3, 4);
     for (size_t i = 0; i < 3; i++) {
       for (size_t j = 0; j < 5; j++)
-        ws->dataX(i)[j] = double(i) + double(j);
+        ws->mutableX(i)[j] = double(i) + double(j);
       for (size_t j = 0; j < 4; j++) {
-        ws->dataY(i)[j] = double(i) + double(j) * 2;
-        ws->dataE(i)[j] = double(i) + double(j) * 3;
+        ws->mutableY(i)[j] = double(i) + double(j) * 2;
```
Comment (LamarMoore) [inline MantidQt/API/test/QwtWorkspaceSpectrumDataTest.h]: See previous comments

File: MantidQt/CustomInterfaces/src/DynamicPDF/DPDFFourierTransform.cpp [hunk 4]
```diff
@@ -95,11 +94,11 @@ void FourierTransform::extractResidualsHistogram(
...
-    residualsWorkspace->dataX(0) = modelWorkspace->dataX(0);
-    residualsWorkspace->dataY(0) =
-        modelWorkspace->dataY(2); // residuals is the third spectrum
-    residualsWorkspace->dataE(0) =
-        modelWorkspace->dataE(0); // errors are coming from experiment
+    residualsWorkspace->setSharedX(0, modelWorkspace->sharedX(0));
+    residualsWorkspace->setSharedY(
+        0, modelWorkspace->sharedY(2)); // residuals is the third spectrum
+    residualsWorkspace->setSharedE(
+        0, modelWorkspace->sharedE(0)); // errors are coming from experiment
```
Comment (LamarMoore) [inline MantidQt/CustomInterfaces/src/DynamicPDF/DPDFFourierTransform.cpp]: Just use `residualsWorkspace->setHistogram(0, modelWorkspace->histogram(0));`

File: MantidQt/CustomInterfaces/src/MultiDatasetFit/MDFDatasetPlotData.cpp [hunk 5]
```diff
@@ -85,47 +83,41 @@ DatasetPlotData::~DatasetPlotData() {
...
-  if (ws->isHistogramData()) {
-    auto xend = xValues.end() - 1;
-    for (auto x = xValues.begin(); x != xend; ++x) {
-      *x = (*x + *(x + 1)) / 2;
-    }
-    xValues.pop_back();
-  }
-  m_dataCurve->setData(xValues.data(), ws->readY(wsIndex).data(),
-                       static_cast<int>(xValues.size()));
+  const auto &xVals = ws->isHistogramData() ? ws->points(wsIndex).rawData()
```
Comment (LamarMoore) [inline MantidQt/CustomInterfaces/src/MultiDatasetFit/MDFDatasetPlotData.cpp]: Why not just always use points. Above, points are being evaluated anyways.

File: MantidQt/CustomInterfaces/src/DynamicPDF/DPDFFourierTransform.cpp (hunk 4 above)
Comment (PranavBahuguna) [inline MantidQt/CustomInterfaces/src/DynamicPDF/DPDFFourierTransform.cpp]: Doesn't work here since Y is set from the third spectrum rather than first.

**Comment by LamarMoore** [comment]
:shipit:

File: MantidQt/API/src/QwtWorkspaceSpectrumData.cpp [hunk 6]
```diff
@@ -75,26 +69,17 @@ Return the y value of data point i
 @return y Y value of data point i
 */
 double QwtWorkspaceSpectrumData::getY(size_t i) const {
-  double tmp = i < m_Y.size() ? m_Y[i] : m_Y.back();
-  if (m_isDistribution) {
-    tmp /= fabs(m_X[i + 1] - m_X[i]);
-  }
-  return tmp;
+  return m_isDistribution ? m_hist.frequencies()[i] : m_hist.counts()[i];
```
Comment (SimonHeybrock) [inline MantidQt/API/src/QwtWorkspaceSpectrumData.cpp]: This looks wrong. The `frequencies()` and `counts()` methods deal with conversion between distribution and count data depending on an *internal* state of the histogram. However, the `m_isDistribution` flag is not necessarily reflecting the internal state, since it depends on what the user wants to plot.

File: MantidQt/API/src/QwtWorkspaceSpectrumData.cpp [hunk 7]
```diff
@@ -75,26 +69,17 @@ Return the y value of data point i
...
 }
 
-double QwtWorkspaceSpectrumData::getEX(size_t i) const {
-  return m_isHistogram ? (m_X[i] + m_X[i + 1]) / 2 : m_X[i];
-}
+double QwtWorkspaceSpectrumData::getEX(size_t i) const { return m_hist.x()[i]; }
```
Comment (SimonHeybrock) [inline MantidQt/API/src/QwtWorkspaceSpectrumData.cpp]: See comment on `getY`.

File: MantidQt/API/src/QwtWorkspaceSpectrumData.cpp [hunk 8]
```diff
@@ -75,26 +69,17 @@ Return the y value of data point i
...
 
 double QwtWorkspaceSpectrumData::getE(size_t i) const {
-  double ei = (i < m_E.size()) ? m_E[i] : m_E.back();
-  if (m_isDistribution) {
-    ei /= fabs(m_X[i + 1] - m_X[i]);
-  }
-  return ei;
+  return m_isDistribution ? m_hist.frequencyStandardDeviations()[i]
+                          : m_hist.countStandardDeviations()[i];
```
Comment (SimonHeybrock) [inline MantidQt/API/src/QwtWorkspaceSpectrumData.cpp]: See comment on `getY`.

File: MantidQt/API/src/QwtWorkspaceSpectrumData.cpp [hunk 9]
```diff
@@ -107,7 +92,7 @@ QString QwtWorkspaceSpectrumData::getXAxisLabel() const { return m_xTitle; }
 QString QwtWorkspaceSpectrumData::getYAxisLabel() const { return m_yTitle; }
 
 bool QwtWorkspaceSpectrumData::setAsDistribution(bool on) {
-  m_isDistribution = on && m_isHistogram;
+  m_isDistribution = on;
```
Comment (SimonHeybrock) [inline MantidQt/API/src/QwtWorkspaceSpectrumData.cpp]: Changed behavior?

File: MantidQt/API/test/QwtWorkspaceSpectrumDataTest.h [hunk 10]
```diff
@@ -32,14 +34,11 @@ class QwtWorkspaceSpectrumDataTest : public CxxTest::TestSuite {
       TS_ASSERT_DELTA(data.e(i), offset + double(i) * 3, 1e-5);
     }
     TS_ASSERT_DELTA(data.x(4), offset + double(4), 1e-5);
-    // Go past the end in Y/E is safe
-    TS_ASSERT_DELTA(data.y(4), offset + double(3) * 2, 1e-5);
```
Comment (SimonHeybrock) [inline MantidQt/API/test/QwtWorkspaceSpectrumDataTest.h]: Why was this removed?

File: MantidQt/CustomInterfaces/src/MultiDatasetFit/MDFDatasetPlotData.cpp [hunk 11]
```diff
@@ -85,19 +83,19 @@ DatasetPlotData::~DatasetPlotData() {
...
                               int wsIndex,
                               const Mantid::API::MatrixWorkspace *outputWS) {
   bool haveFitCurves = outputWS && outputWS->getNumberHistograms() >= 3;
-  auto xValues = ws->points(wsIndex);
 
-  m_dataCurve->setData(xValues.rawData().data(),
-                       ws->y(wsIndex).rawData().data(),
-                       static_cast<int>(xValues.size()));
+  const auto &xVals = ws->isHistogramData() ? ws->points(wsIndex).rawData()
+                                            : ws->binEdges(wsIndex).rawData();
```
Comment (SimonHeybrock) [inline MantidQt/CustomInterfaces/src/MultiDatasetFit/MDFDatasetPlotData.cpp]: Changed behavior??

**Review by SimonHeybrock — COMMENTED** [review]
There seem to be a few things wrong here with the mechanism of normalizing data in plots. @mdoucet reported broken plotting.

See my comments:

File: MantidQt/API/src/QwtWorkspaceSpectrumData.cpp [hunk 12]
```diff
@@ -52,21 +52,15 @@ QwtWorkspaceSpectrumData *QwtWorkspaceSpectrumData::copyWithNewSource(
...
+size_t QwtWorkspaceSpectrumData::size() const { return m_hist.size(); }
 
 /**
 Return the x value of data point i
 @param i :: Index
 @return x X value of data point i
 */
 double QwtWorkspaceSpectrumData::getX(size_t i) const {
-  return m_binCentres ? (m_X[i] + m_X[i + 1]) / 2 : m_X[i];
+  return m_binCentres ? m_hist.points()[i] : m_hist.binEdges()[i];
```
Comment (PranavBahuguna) [inline MantidQt/API/src/QwtWorkspaceSpectrumData.cpp]: Is there a way of converting a  whole histogram from bin edges to points in place? I'm aware that you can use `convertToCounts` and `convertToFrequencies` for Y and E, but is there any equivalent for X?

**Comment by martyngigg** [comment]
We need to do something about the broken behaviour here ASAP. Users are starting to notice.

Should we revert this PR? @LamarMoore @SimonHeybrock @PranavBahuguna

**Comment by PranavBahuguna** [comment]
I'm working on fixing this, but I suggest reverting this for now as it can take some time to fix.
